node/src/blockchain/blockchain_bridge.rs

```diff
@@ -232,229 +246,238 @@ impl BlockchainBridge {
+            .collect::<Vec<Hash>>();
+
+        Box::new(
+            self.blockchain_interface
+                .lower_interface()
+                .get_transaction_receipt_batch(transaction_hashes)
+                .map_err(move |e| e.to_string())
+                .and_then(move |transaction_receipts_results| {
+                    let length = transaction_receipts_results.len();
+                    let mut transactions_found = 0;
+                    for transaction_receipt in &transaction_receipts_results {
+                        if let TransactionReceiptResult::Found(_) = transaction_receipt {
+                            transactions_found += 1;
+                        }
+                    }
+                    let pairs = transaction_receipts_results
+                        .into_iter()
+                        .zip(msg.pending_payable.into_iter())
+                        .collect_vec();
+                    accountant_recipient
+                        .try_send(ReportTransactionReceipts {
+                            fingerprints_with_receipts: pairs,
+                            response_skeleton_opt: msg.response_skeleton_opt,
+                        })
+                        .expect("Accountant is dead");
+                    if length != transactions_found {
+                        return Err(format!(
+                            "Aborting scanning; {} transactions succeed and {} transactions failed",
+                            transactions_found,
+                            length - transactions_found
+                        ));
+                    };
+                    Ok(())
+                }),
+        )
     }
 
-    fn handle_scan<M, F>(&mut self, handler: F, scan_type: ScanType, msg: M)
+    fn handle_scan_future<M, F>(&mut self, handler: F, scan_type: ScanType, msg: M)
     where
-        F: FnOnce(&mut BlockchainBridge, M) -> Result<(), String>,
+        F: FnOnce(&mut BlockchainBridge, M) -> Box<dyn Future<Item = (), Error = String>>,
         M: SkeletonOptHolder,
     {
         let skeleton_opt = msg.skeleton_opt();
-        match handler(self, msg) {
-            Ok(_r) => (),
-            Err(e) => {
-                warning!(self.logger, "{}", e);
-                self.scan_error_subs_opt
-                    .as_ref()
-                    .expect("Accountant not bound")
-                    .try_send(ScanError {
-                        scan_type,
-                        response_skeleton_opt: skeleton_opt,
-                        msg: e,
-                    })
-                    .expect("Accountant is dead");
-            }
-        }
+        let logger = self.logger.clone();
+        let scan_error_subs_opt = self.scan_error_subs_opt.clone();
+        let future = handler(self, msg).map_err(move |e| {
+            warning!(logger, "{}", e);
+            // TODO: This ScanError needs to be removed, And added into OutboundPaymentsInstructions & QualifiedPayablesMessage
```

Comment:
**URGENCY STATUS: I DON'T WANT TO LET IN CONFUSING COMMENTS, SO PLEASE SET YOUR EYES AT THIS.

Reply:
We have updated this comment to explain what our thinking was.
Please let us know your thoughts.

```rust
// TODO: This ScanError needs to be removed, And added into OutboundPaymentsInstructions & QualifiedPayablesMessage
            //   There are certain cases when its a partial error and we are triggering errors that will send ScanError messages.
            //   In case we dont send this message at all and instead we use the above two mentioned messages to send total failure and partial failure.
            //   BlockchainBridge wont segregate the messages and Accountant can later on deal with success, partial failures and total failures accordingly.
```

Reply:
Eh, this comment didn't help me much with understanding your view of things

I have two points about what concerns me.

a) Regardless what you mean by "partial errors/failures", because that's what is totally blur to me and unless it's explained I don't think this comment might do any good, I really have a problem with something which seems to be against common sense. Both of those cited messages go FROM Accountant TO BlockchainBridge.
Only Accountant is waiting in this process, it's not the BlockchainBridge. So, if you say you've caught an error, which must have happened within Accountant's message handler, why do you bother sending any additional message to the BlockchainBridge despite it would be so convenient to halt it right away once you know about an error. In other words, any error carrying in the direction outwards the Accountant lacks a good reason to me.
If you think there is a reason, or if you think that you HAVE TO operate with some kind of partial failures, it rather smells with a bad design. Please, think this through again carefully. 

b) There were times, before you came to modify the code for futures specifically, when we had simply a `handle_scan` method. I liked that it was a shared method among all those three scanners with their own procedures. Basically, the concept has ever been like we start in Accountant, send some task to BlockchainBridge, and accepts a response back in Accountant. 
This `handle_scan` method was to embrace each of the main methods in the BlockchainBridge responsible for handling a particular scan. It didn't matter whether it was for Payables, PendingPayables or Receivables. 
Still, it always was ready to take over an error from the specialized scan method and process it the same way in all three cases. It was fine when this procedure established as a logging, and grabbing the error and afterwards also sending the ScanError message with an identical message to the one just logged.
(I don't want to cause this explanation to become overly complex but I think that with the birth of the BlockchainAgent, and the split of the Payable scan into four instead of two messages exchange, the `handle_scan` probably turned to be employed four times, even though the scanners are only three).

After what I just said, you can see with me there is something wrong, because your comment speaks only about the Payable scanner to which the two messages you mention belong to. Therefore, are you really sure, that you want to deal with each scanner individually, laboriously, instead of leaving the ScanError sending in this single place to be dispatched in a standardized manner for each the scan processes?